lnwire/message_test.go

```diff
@@ -0,0 +1,844 @@
+	"sync"
+	"testing"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcutil"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/tor"
+	"github.com/stretchr/testify/require"
+)
+
+const deliveryAddressMaxSize = 34
+const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+var (
+	testSig = &btcec.Signature{
+		R: new(big.Int),
+		S: new(big.Int),
+	}
+	testNodeSig, _ = lnwire.NewSigFromSignature(testSig)
+
+	testNumExtraBytes = 1000
+	testNumSigs       = 100
+	testNumChanIDs    = 1000
+	buffer            = make([]byte, 0, lnwire.MaxSliceLength)
+
+	bufPool = sync.Pool{
+		New: func() interface{} {
+			return bytes.NewBuffer(buffer)
+		},
+	}
+)
+
+// BenchmarkWriteMessage benchmarks the performance of lnwire.WriteMessage. It
+// generates a test message for each of the lnwire.Message, calls the
+// WriteMessage method and benchmark it.
+func BenchmarkWriteMessage(b *testing.B) {
+	// Create testing messages. We will use a constant seed to make sure
+	// the benchmark uses the same data every time.
+	r := rand.New(rand.NewSource(42))
+
+	msgAll := makeAllMessages(b, r)
+
+	// Iterate all messages and write each once.
+	for _, msg := range msgAll {
+		m := msg
+		// Run each message as a sub benchmark test.
+		b.Run(msg.MsgType().String(), func(b *testing.B) {
+
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				// Fetch a buffer from the pool and reset it.
+				buf := bufPool.Get().(*bytes.Buffer)
+				buf.Reset()
+
+				_, err := lnwire.WriteMessage(buf, m, 0)
+				require.NoError(b, err, "unable to write msg")
+
+				// Put the buffer back when done.
+				bufPool.Put(buf)
+			}
+		})
+	}
+}
+
+// BenchmarkReadMessage benchmarks the performance of lnwire.ReadMessage. It
+// first creates a test message for each of the lnwire.Message, writes it to
+// the buffer, then later reads it from the buffer.
+func BenchmarkReadMessage(b *testing.B) {
+	// Create testing messages. We will use a constant seed to make sure
+	// the benchmark uses the same data every time.
+	r := rand.New(rand.NewSource(42))
+	msgAll := makeAllMessages(b, r)
+
+	// Write all the messages to the buffer.
+	for _, msg := range msgAll {
+		// Fetch a buffer from the pool and reset it.
+		buf := bufPool.Get().(*bytes.Buffer)
+		buf.Reset()
+
+		_, err := lnwire.WriteMessage(buf, msg, 0)
+		require.NoError(b, err, "unable to write msg")
+
+		// Run each message as a sub benchmark test.
+		m := msg
+		b.Run(m.MsgType().String(), func(b *testing.B) {
+			b.ResetTimer()
+
+			for i := 0; i < b.N; i++ {
+				r := bytes.NewBuffer(buf.Bytes())
+
+				// Read the message from the buffer.
+
+				_, err := lnwire.ReadMessage(r, 0)
+				require.NoError(b, err, "unable to read msg")
+			}
+		})
+
+		// Put the buffer back when done.
+		bufPool.Put(buf)
+	}
+}
+
+// makeAllMessages is used to create testing messages for each lnwire message
+// type.
+//
+// TODO(yy): the following testing messages are created somewhat arbitrary. We
```

Comment:
What do you mean by standardize? Like set the set of fields and TLV values that we typically see in practice?

Reply:
Yeah exactly. Atm I'm just filling random bytes into each field, while it's a good start, we may someday get to the point where we want to optimize a single byte (hopefully!).